documentation/docs/explanations/deadlock-analysis.mdx

```diff
@@ -0,0 +1,179 @@
+
+##### Analysis
+
+The graph at the right represents the merged dependency graph of functions $f_1$ and $f_2$. It shows all the dependencies of the locks in $f_1$ and $f_2$.
+The presence of a circular dependency on the merged graph indicates the possibility of a deadlock when executing $f_1$ and $f_2$ concurrently.
+
+##### Example run resulting in a deadlock: 
+1. $f_1$ starts and takes the lock $1$
+2. $f_2$ starts and takes the lock $2$
+3. $f_1$ is stuck because $f_2$ is holding the lock $2$ that it requires
+4. $f_2$ is stuck because $f_2$ is holding the lock $1$ that it requires
+
+
+
+#### Example: Deadlock with 3 functions
+![](assets/deadlock_3_functions.drawio.png)
+
+As the previous example there is a circular dependency of locks 
+between $f_1$, $f_2$ and $f_3$. 
+Example run resulting in a deadlock: 
+1. $f_1$ starts and takes the lock $1$
+2. $f_2$ starts and takes the lock $3$
+3. $f_3$ starts and takes the lock $2$
+4. $f_1$ is stuck because $f_3$ is holding the lock $2$ that it requires
+5. $f_2$ is stuck because $f_1$ is holding the lock $1$ that it requires
+6. $f_3$ is stuck because $f_2$ is holding the lock $3$ that it requires
+
+### Proof backing "_no circular dependency equals to no deadlock_"
+This is a proof on an example for simplicity. The reasoning used in the next part can be applied to any dependency graph of locks to prove the absence of deadlock when there are no circular dependency.
+
+#### Example: no deadlock
+![](assets/no_deadlock.drawio.png)
+
+There is no circular dependency in the right-side representation.
+We will try to argue that there is indeed no possible deadlock in this configuration.
+The main idea linking absence of deadlock with absence of circular dependency is that any thread will always eventually get the lock it's waiting for. Thus it will eventually terminate its computation thus releasing the locks it acquired.
+
+##### Definition: deadlock-free lock
+We say a lock is deadlock-free when we can prove that it will never part of a deadlock and will always eventually be released.
+
+#### Proof of the absence of deadlock using the merged graph
+
+1. __Lock $3$ is deadlock-free__. No locks depend on lock $3$. When a thread acquires lock $3$ it will manage to terminate its computation and will always release it
+2. __Recursively lock $2$ is deadlock-free__. Lock $2$ has only lock $3$ which depends on it. That means:
+  a) if a thread acquires lock $2$ it will eventually get lock $3$ because lock $3$ is deadlock-free (lock $3$ will always be available at some point)
+  b) if a thread acquires lock $2$ it will manage to terminate its computation and eventually be released because lock $3$ is deadlock-free
+  c) lock $2$ is deadlock-free by recursivity
+3. __Recursively lock $1$ is deadlock-free__. All the locks that depend on lock $1$ are deadlock-free. Reasoning for lock $2$ can be also applied to lock $1$.
```

Comment:
```suggestion
#### Prove the absence of deadlock using the merged graph

1. __Lock $3$ is deadlock-free__. No locks depend on lock $3$. When a thread acquires lock $3$ it will manage to terminate its computation and will always release it.
2. __Recursively lock $2$ is deadlock-free__. Lock $3$ is the only lock that depends on lock $2$. That means:
    1. If a thread acquires lock $2$, it will eventually get lock $3$ because lock $3$ is deadlock-free (lock $3$ will always be available at some point).
    2. If a thread acquires lock $2$, it will manage to terminate its computation and eventually be released because lock $3$ is deadlock-free.
    3. Lock $2$ is deadlock-free by recursivity.
3. __Recursively, lock $1$ is deadlock-free__. All the locks that depend on lock $1$ are deadlock-free. You can apply the same reasoning for lock $2$ to lock $1$.
```